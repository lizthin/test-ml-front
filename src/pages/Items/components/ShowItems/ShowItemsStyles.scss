@import "../../../../assets/styles/colors.scss";
@import "../../../../assets/styles/mixins.scss";

.left-item {
    float: left;
}

.card-items {
    background-color: $white;
    cursor: pointer;
    @include for-phone-only {
        margin-top: 16px;
    }
}

.card-items:hover {
    background-color: $gray-light-hover;
    cursor: pointer;
}

.container-img-card-item {
    width: 180px;
    height: 180px;
    border-radius: 4px;
    float: left;
    margin: 16px;
    @include for-phone-only {
        width: 70%;
        height: 50%;
        border-radius: 4px;
        float: left;
        margin: 35px 58px 0px;
    }
}

.container-detail-card-item {
    float: left;
    width: 336px;
    @include for-phone-only {
        padding-left: 16px;
        width: 100%;
    }
}

.price-card-items {
    font-size: 24px;
    color: #333;
    float: left;
    width: 100%;
}

.title-card-items {
    white-space: pre-wrap;
    font-size: 18px;
    color: #333333;
    float: left;
    width: 100%;
}

.container-img-shipping {
    width: 17px;
    height: 17px;
    float: left;
    margin-left: 6px;
    margin-top: -1px;
}

.currency-card-item {
    font-size: 12px;
    color: $gray-dark;
    margin-top: 10%;
    margin-left: 85%;
    position: absolute;
    @include for-phone-only {
        position: inherit;
        margin-left: 5%;
    }
}

.currency-container-card-item {
    @include for-phone-only {
        position: absolute;
        margin-top: 267px;
        margin-left: 284px;
    }
    @include for-tablet-portrait-up {
        margin-top: -215px;
        margin-left: 566px;
    
    }
}

hr {
    margin-block-end: 0;
    margin-block-start: 0;
    margin-inline-end: 0;
    margin-inline-start: 0;
    color: $gray-light;
    width: 95%;
    margin-left: 2%;
    border-bottom: 1px solid $gray-light;
    border-top: 0;
    border-left: 0;
    border-right: 0;
}

.img-card-item,
.img-shipping {
    width: 100%;
    height: 100%;
}