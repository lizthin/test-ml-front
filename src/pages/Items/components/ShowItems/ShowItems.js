import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import ItemsServices from "../../../../services/ItemService/ItemsService";
import "./ShowItemsStyles.scss";
import freeShipping from "../../../../assets/img/icons/ic_shipping@2x.png.png.png";
import Breadcrumb from "../../../../components/Breadcrumb/Breadcrumb";

/**
 * funtion component Show Items, make list items input search response api
 * @param {object} props history router props, data items initial state items  
 * @returns list items, breadcrumb catefories
 */
const ShowItems = ({ history, dataItems }) => {

  /**
   * Hook useState items, search list items
   * @example setItems({id:'ejemplo', title: 'ejemplo', price: {}},{id:'ejemplo', title: 'ejemplo', price: {}})
   */
  const [items, setItems] = useState(dataItems ?? []);

  /**
   * Hook useState categories, breadCrumb output
   * @example setCategories(['bolsos', 'maquillaje'])
   */
  const [categories, setCategories] = useState([]);

   /**
   * Hook useEffect call service fetchAllItemsFiltered response array items
   * activate with change location search
   */
  useEffect(() => {
    if (history.location.search !== "") {
      const query = history.location.search.split("=")[1];

      ItemsServices.fetchAllItemsFiltered(query)
        .then(({ items, categories }) => {
          setItems(items.slice(0, 4));
          setCategories(categories.slice(0, 5));
        })
        .catch((error) => console.warn);
    }
  }, [history.location.search]);

  /**
   * Funtion handleRedirectUrlDetail, change histori props pathname 
   * activate ShowDetail component
   * @param {string} id item details
   * @example handleRedirectUrlDetail('MA12154') 
   */
  const handleRedirectUrlDetail = (id) => {
    history.push({
      pathname: `/items/${id.trim()}`,
      data: {
        categories: categories,
      },
    });
  };

  /**
   * const shipping put class free shiping item
   */
  const shipping = (
    <div className="container-img-shipping">
      <img src={freeShipping} alt={freeShipping} className="img-shipping" />
    </div>
  );

  return (
    <div className="container">
      <div className="row justify-content-around">
        <div className="col-10">
          <div className="row">
            <div className="col-12">
              {categories.length > 0 ? (
                categories.map((category, index) => (
                  <Breadcrumb
                    key={category}
                    properties={{
                      category,
                      lengthCatgories: categories.length,
                      index,
                      history,
                    }}
                  />
                ))
              ) : (
                <>
                  <div className="row mt-2 pt-2"></div>
                </>
              )}
            </div>
          </div>
          {items.map((item) => (
            <div className="row justify-content-around" key={item.id}>
              <div
                className="col-12 card-items"
                onClick={() => {
                  handleRedirectUrlDetail(item.id);
                }}
              >
                <div className="row">
                  <div className="col-12">
                    <div className="container-img-card-item">
                      <img
                        className="img-card-item"
                        src={item.picture}
                        alt={item.picture}
                      />
                    </div>
                    <div className="currency-card-item">{item.city}</div>
                    <div className="container-detail-card-item">
                    
                      <div className="price-card-items top-32-general">
                        <div className="left-item">
                          ${" "}
                          {new Intl.NumberFormat("de-DE").format(
                            item.price.amount
                          )}
                        </div>
                        {item.free_shipping ? shipping : <></>}
                      </div>
                      <p className="title-card-items">{item.title}</p>
                    </div>
                  </div>
                </div>
                <hr />
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

ShowItems.propTypes = {
  history: PropTypes.object.isRequired,
};

export default ShowItems;