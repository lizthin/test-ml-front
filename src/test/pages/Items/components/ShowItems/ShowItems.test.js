import React from "react";
import { shallow, mount } from "enzyme";
import ShowItems from "../../../../../pages/Items/components/ShowItems/ShowItems";
import { act } from "@testing-library/react";
import ItemsServices from "../../../../../services/ItemService/ItemsService";

describe("should show items list with querty", () => {
  const mockProperties = {
    location: {
      hash: "",
      key: "6ti0yy",
      pathname: "/items",
      search: "?search=futbol",
    },
    push: jest.fn()
  };

  const mockItems = [{
    id: "MLA863083993",
    title: "Peluche Chimuelo / Furia Luminosa 30 Cm Hermosos",
    price: {
      currency: "ARS",
      amount: 2190,
      decimals: 0,
    },
    picture: "http://http2.mlstatic.com/D_843845-MLA42260331923_062020-I.jpg",
    condition: "new",
    free_shipping: true,
    sold_quantity: 25,
    description:
      "LEER PRECIO X CADA UNO \n\nHermoso chimuelo - furia luminosa de entrenando a tu dragón 30 cm hermosos \nPRECIO X CADA UNO\n-Entregamos a una cuadra de SAN JUAN Y BOEDO\n\n-HACEMOS ENVIOS en caba (Se realiza por motomensajería, si realiza la compra antes de las 12 m de lunes a viernes en días hábiles la entrega se realiza en el día, si realiza la compra después de las 12 am la entrega se realiza el primer día hábil luego de realizada la compra, si necesita que realicemos la entrega en una franja horaria al finalizar la compra escríbanos su preferencia por mensajería de Mercadolibre para tenerlo en cuenta.\n\n-Servicio de Mensajería en CABA , consulte el costo, los días de lluvia el servicio tiene un recargo del 50%\n\n-ENVIOS A TODO EL PAIS (Realizamos envíos hasta su domicilio dentro de la República Argentina por el servicio de MercadoEnvíos. El valor del mismo lo obtendrá ingresando su código postal en el calculador.\n\n-SU PREGUNTA NO MOLESTA",
  }];

  const mockCategories = ['peluches'];

  test("should component with item.free_shipping true", () => {
    const wrapper = shallow(<ShowItems history={mockProperties} dataItems={mockItems}/>);
    wrapper.find('.card-items').simulate('click')
    expect(wrapper).toMatchSnapshot()
  });

  test("should component with item.free_shipping false", () => {
    mockItems[0].free_shipping = false;
    const wrapper = shallow(<ShowItems history={mockProperties} dataItems={mockItems}/>);
    wrapper.find('.card-items').simulate('click')
    expect(wrapper).toMatchSnapshot()
  });

  test("should do useEffect with service ItemsServices result", async () => {
    await act(async () => {
      jest.spyOn(ItemsServices, 'fetchAllItemsFiltered').mockResolvedValue({ items: mockItems, categories: mockCategories });

      const wrapper = mount(<ShowItems history={mockProperties} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  test("should useEffect with service ItemsServices error", async () => {
    await act(async () => {
      jest.spyOn(ItemsServices, 'fetchAllItemsFiltered').mockResolvedValue(undefined);
      const wrapper = mount(<ShowItems history={mockProperties} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  test("should useEffect with service and mockProperties.location.search empty", async () => {
    await act(async () => {
      jest.spyOn(ItemsServices, 'fetchAllItemsFiltered').mockResolvedValue(undefined);
      mockProperties.location.search = '';
      const wrapper = mount(<ShowItems history={mockProperties} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

});