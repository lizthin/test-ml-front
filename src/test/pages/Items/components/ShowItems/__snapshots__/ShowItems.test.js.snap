// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show items list with querty should component with item.free_shipping false 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-around"
  >
    <div
      className="col-10"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="row mt-2 pt-2"
          />
        </div>
      </div>
      <div
        className="row justify-content-around"
        key="MLA863083993"
      >
        <div
          className="col-12 card-items"
          onClick={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="col-12"
            >
              <div
                className="container-img-card-item"
              >
                <img
                  alt="http://http2.mlstatic.com/D_843845-MLA42260331923_062020-I.jpg"
                  className="img-card-item"
                  src="http://http2.mlstatic.com/D_843845-MLA42260331923_062020-I.jpg"
                />
              </div>
              <div
                className="currency-card-item"
              />
              <div
                className="container-detail-card-item"
              >
                <div
                  className="price-card-items top-32-general"
                >
                  <div
                    className="left-item"
                  >
                    $
                     
                    2,190
                  </div>
                </div>
                <p
                  className="title-card-items"
                >
                  Peluche Chimuelo / Furia Luminosa 30 Cm Hermosos
                </p>
              </div>
            </div>
          </div>
          <hr />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should show items list with querty should component with item.free_shipping true 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-around"
  >
    <div
      className="col-10"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="row mt-2 pt-2"
          />
        </div>
      </div>
      <div
        className="row justify-content-around"
        key="MLA863083993"
      >
        <div
          className="col-12 card-items"
          onClick={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="col-12"
            >
              <div
                className="container-img-card-item"
              >
                <img
                  alt="http://http2.mlstatic.com/D_843845-MLA42260331923_062020-I.jpg"
                  className="img-card-item"
                  src="http://http2.mlstatic.com/D_843845-MLA42260331923_062020-I.jpg"
                />
              </div>
              <div
                className="currency-card-item"
              />
              <div
                className="container-detail-card-item"
              >
                <div
                  className="price-card-items top-32-general"
                >
                  <div
                    className="left-item"
                  >
                    $
                     
                    2,190
                  </div>
                  <div
                    className="container-img-shipping"
                  >
                    <img
                      alt="ic_shipping@2x.png.png.png"
                      className="img-shipping"
                      src="ic_shipping@2x.png.png.png"
                    />
                  </div>
                </div>
                <p
                  className="title-card-items"
                >
                  Peluche Chimuelo / Furia Luminosa 30 Cm Hermosos
                </p>
              </div>
            </div>
          </div>
          <hr />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should show items list with querty should do useEffect with service ItemsServices result 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-around"
  >
    <div
      className="col-10"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="row mt-2 pt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should show items list with querty should useEffect with service ItemsServices error 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-around"
  >
    <div
      className="col-10"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="row mt-2 pt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should show items list with querty should useEffect with service and mockProperties.location.search empty 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-around"
  >
    <div
      className="col-10"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="row mt-2 pt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
